---
title: "Getting started with the pseudo-bulk simulator R package"
bibliography: references.bib
biblio-style: apalike
link-citations: yes
colorlinks: yes
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{simulator_documentation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(simulator)
library(ggplot2)
library(RColorBrewer)
```

# Introduction
Cell-type deconvolution is an important analysis-step while processing bulk RNA-seq experiments 
and many tools exist to tackle this issue. Though the user lacks a good comparison between those
tools and when one might be more beneficial than others. \
The goal of this simulator is to create benchmark datasets, where pseudo bulk sample with 
predefined cell-type fractions can be simulated using expression data from single-cell RNAseq
experiments. 


# Setup the simulator
This package can simulate samples from local or public data. We will start with the public data
integration, since it requires no mostly no manual download of datasets. As a public database, 
sfaira [@Fischer2020] is used. This gives access to 233 datasets from human and mouse with 
more than 3 million cells in total. \
In order to use this database, we first need to install it using conda. Simply follow [this installation
tutorial](https://sfaira.readthedocs.io/en/latest/installation.html). You do not need to activate the
new conda environment in order to use it later on. \
To setup sfaira inside of the R package use this command: 

```{r}
setup_list <- simulator::setup_sfaira(python_path = "/nfs/home/students/adietrich/.conda/envs/sfaira/bin/python3",
                                      env_name = "sfaira", 
                                      basedir = "/nfs/home/students/adietrich/ma/data/sfaira")
```

The `python_path` parameter is used to set the path to the python executable **inside** of the
newly created conda environment. Similarily, the `env_name` parameter is the name of this newly
created conda environment. These two parameters need to be set correctly, so that sfaira can be
used later on. \
The third parameter describes the name of a direcory in which the raw files, meta data and cached
data from downloading datasets from sfaira are stored into. \
You need to save the output of the command above in a variable for later use. \
If you do not wish to use sfaira, just skip this step entirely. 

# Creating a dataset

The simulator package works with two internally defined data-structures: *datasets* and *databases*. \
We will now create a dataset of samples from human pancreas:

```{r}
ds_human_pancreas <- simulator::dataset_sfaira_multiple(sfaira_setup = setup_list,
                                                        organisms = "human", 
                                                        tissues = "pancreas", 
                                                        name="human_pancreas")
```

This will download 2 datasets from sfaira and merge them together into a single expression
matrix and a streamlined annotation table, which we can use for our simulation.

# Simulate pseudo bulk datasets

We are now ready to simulate the first pseudo bulk samples with the created dataset:

```{r}
simulation <- simulator::simulate_bulk(data = ds_human_pancreas,
                                       scenario = "random", 
                                       scaling_factor = "none", 
                                       ncells=1000, 
                                       nsamples = 50, 
                                       ncores = 12)
```

`ncells` sets the number of cells in each sample, while `nsamples` sets the total amount
of simulated samples. \
You can also control the number of cells per sample by using the `total_read_counts`
parameter, which sets the total number of counts per sample. This can be used to simulate different
sequencing depths. If you wish to use this parameter, it is important to know what type of
expression data you have (raw or normalized). If you for example use TPMs as expression values
and set the total read counts to `1e8`, this will get you really huge samples with many cells, since
TPM values are generally smaller than the raw count values. 

Currently there are 4 `scenarios` implemented in the package: \
 
* *random*: for each cell in a sample, the cell-type is chosen randomly. This scenario is 
depending on the background distribution of the cell-types in the created dataset. If for example
your dataset contains 20% B-cells, 40% T-cells and 40% Macrophages, this scenario will create samples
where this distribution is mirrored closely. \

* *uniform*: this creates samples, where all existing cell-types in the dataset appear in 
the same proportions. So using a dataset with 3 cell-types, this will simulate samples, where
all cell-type fractions are 1/3. \

* *spike-in*: here you need to set two additional parameters for the `simulate_bulk()` function:
`spike_in_cell_type` sets the cell-type you want to be over-representing and 
`spike_in_amount` sets the fraction of this cell-type. You could for example use `B-cell` and `0.5`
to create samples, where 50% are B-cells and the rest is filled randomly with other cell-types. \

* *spill-over*: this simulates samples with only a single cell-type present. Like with *spike-in*
you need to set this cell-type with the additional parameter `spillover_cell_type`. \

# Results

The `simulation` object contains three named entries: \

* `pseudo_bulk`: a sparse matrix with - in this case - 50 columns and about 60k rows 
(this is the number of features/genes). This is the simulated expression matrix. \

* `cell_fractions`: a dataframe with 50 rows and 15 columns (this is the number of cell-types).
It gives the fractions for each cell-type in each sample. \

* `expression_set`: a Biobase [ExpressionSet](https://rdrr.io/bioc/Biobase/man/class.ExpressionSet.html) \

Finally here is a boxplot of the resulting simulation:

```{r}
fractions <- simulation$cell_fractions
fractions$sample <- rownames(frac)
frac_long <- gather(fractions, cell_type, fraction, 1:length(fractions)-1)

ggplot(frac_long, aes(x=fraction, y=sample, fill=cell_type))+
  geom_col()+
  ggtitle("Cell-type fractions for 50 pseudo-bulk RNAseq samples")+
  scale_fill_manual(values = colorRampPalette(brewer.pal(12, "Set2"))(length(unique(frac_long$cell_type))))
```




# References
