---
title: "Getting started with the pseudo-bulk simulator R package"
author: "Alexander Dietrich"
bibliography: references.bib
biblio-style: apalike
link-citations: yes
colorlinks: yes
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{simulator_documentation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Installation

To install the package, run this:
(Since it is a private repository, you currently need an authentification token for this repository.)

```{r, eval=FALSE}
install.packages("devtools")
devtools::install_github("omnideconv/simulator") #add auth=XXX to download from private repository
```


```{r setup}
library(simulator)
library(data.table)
library(tidyr)
library(Matrix)
library(Seurat)
library(ggplot2)
library(RColorBrewer)
```

# Introduction
Cell-type deconvolution is an important analysis-step while processing bulk RNA-seq experiments 
and many tools exist to tackle this issue. Though the user lacks a good comparison between those
tools and when one might be more beneficial than others. \
The goal of this simulator is to create benchmark datasets, where pseudo bulk samples with 
predefined cell-type fractions can be simulated using expression data from single-cell RNAseq
experiments. 

# Getting started

This chapter covers all you need to know to quickly simulate some pseudo-bulk samples!

## Setup the simulator
This package can simulate samples from local or public data. We will start with the public data
integration, since it requires mostly no manual download of datasets. As a public database, 
sfaira [@Fischer2020] is used, which is a dataset and model repository for single-cell RNA-sequencing
data. It gives access to about 233 datasets from human and mouse with more than 3 million cells in total. 
You can browse them interactively here: [https://theislab.github.io/sfaira-portal/Datasets](https://theislab.github.io/sfaira-portal/Datasets). Note 
that only annotated datasets will be downloaded! \
In order to use this database, we first need to install it. We recommend to use [conda](https://docs.conda.io/projects/conda/en/latest/index.html) to create a new
environment and install sfaira into this directory. This ensures that no other packages interfere with the installation. \
*Note:* currently the simulator seems to work only with the develop version of sfaira.\
A recommended step by step installation would look like this:
```
conda create --name sfaira
conda activate sfaira
conda install pip
git clone https://github.com/theislab/sfaira.git
cd sfaira
git checkout dev
git pull
/path/to/conda/envs/sfaira/bin/pip install -e . #this ensures that the correct pip is used
```
Watch out that you are using pip inside of the new environment, not the global pip!\

You do not need to activate the new conda environment each time in order to use it in the R package.\
To setup sfaira inside of the R package use this command: 

```{r, warning=TRUE, message=TRUE}
setup_list <- simulator::setup_sfaira(python_path = "/nfs/home/students/adietrich/.conda/envs/sfaira/bin/python3",
                                      env_name = "sfaira", 
                                      basedir = "/nfs/home/students/adietrich/ma/data/sfaira")
```

The `python_path` parameter is used to set the path to the python executable **inside** of the
newly created conda environment. Similarly, the `env_name` parameter is the name of this newly
created conda environment. These two parameters need to be set correctly, so that sfaira can be
used later on. \
The third parameter describes the name of a directory in which the raw files, meta data and cached
data from downloading datasets from sfaira are stored into. \
You need to save the output of the command above in a variable for later use. \
If you do not wish to use sfaira, just skip this step entirely. 

## Creating a dataset

The simulator package works with two internally defined data-structures: *datasets* and *databases*. \
We will now create a dataset of samples from human pancreas using the `organisms` and `tissues` parameter.
You can provide a single word (like we do here) or for example a list of tissues
you want to download: `c("pancreas","lung")`. An additional parameter is the `assays` parameter,
where you subset the database further to only download datasets from certain sequencing
assays (for examples `Smart-seq2`). \
The `name` parameter is used later on to give each sample a unique name.

```{r, warning=TRUE, message=TRUE}
ds_human_pancreas <- simulator::dataset_sfaira_multiple(sfaira_setup = setup_list,
                                                        organisms = "human", 
                                                        tissues = "pancreas", 
                                                        name="human_pancreas")
```
Currently there are two datasets in sfaira from human pancreas, which have cell-type annotation.
The package will download them for you automatically and merge them together into a single expression
matrix and a streamlined annotation table, which we can use for our simulation. \
It can happen, that some datasets from sfaira are not (yet) ready for the automatic download,
an error message will then appear in R, telling you which file to download and where to put it. \

## Simulate pseudo bulk datasets

We are now ready to simulate the first pseudo bulk samples with the created dataset:

```{r}
simulation <- simulator::simulate_bulk(data = ds_human_pancreas,
                                       scenario = "random", 
                                       scaling_factor = "none", 
                                       ncells=1000, 
                                       nsamples = 50, 
                                       ncores = 12)
```

`ncells` sets the number of cells in each sample, while `nsamples` sets the total amount
of simulated samples. \
You can also control the number of cells per sample by using the `total_read_counts`
parameter, which sets the total number of counts per sample. This can be used to simulate different
sequencing depths. If you wish to use this parameter, it is important to know what type of
expression data you have (raw or normalized). If you for example use TPMs as expression values
and set the total read counts to `1e7`, this will get you really huge samples with many cells, since
TPM values are generally smaller than the raw count values. 

Currently there are 4 `scenarios` implemented in the package: \
 
* *random*: for each cell in a sample, the cell-type is chosen randomly. This scenario is 
depending on the background distribution of the cell-types in the created dataset. If for example
your dataset contains 20% B-cells, 40% T-cells and 40% Macrophages, this scenario will create samples
where this distribution is mirrored closely. \

* *uniform*: this creates samples, where all existing cell-types in the dataset appear in 
the same proportions. So using a dataset with 3 cell-types, this will simulate samples, where
all cell-type fractions are 1/3. \

* *spike-in*: here you need to set two additional parameters for the `simulate_bulk()` function:
`spike_in_cell_type` sets the cell-type you want to be over-representing and 
`spike_in_amount` sets the fraction of this cell-type. You could for example use `B-cell` and `0.5`
to create samples, where 50% are B-cells and the rest is filled randomly with other cell-types. \

* *spill-over*: this simulates samples with only a single cell-type present. Like with *spike-in*
you need to set this cell-type with the additional parameter `spillover_cell_type`. \

## Results

The `simulation` object contains three named entries: \

* `pseudo_bulk`: a sparse matrix with - in this case - 50 columns and about 60k rows 
(this is the number of features/genes). This is the simulated expression matrix. \

* `cell_fractions`: a dataframe with 50 rows and 15 columns (this is the number of cell-types).
It gives the fractions for each cell-type in each sample. \

* `expression_set`: a Biobase [ExpressionSet](https://rdrr.io/bioc/Biobase/man/class.ExpressionSet.html) \

Finally here is a boxplot of the resulting simulation:

```{r, fig.width=8, fig.height=8}
fractions <- simulation$cell_fractions
fractions$sample <- rownames(fractions)
frac_long <- gather(fractions, cell_type, fraction, 1:length(fractions)-1)

ggplot(frac_long, aes(x=fraction, y=sample, fill=cell_type))+
  geom_col()+
  ggtitle("Cell-type fractions for 50 pseudo-bulk RNAseq samples")+
  scale_fill_manual(values = colorRampPalette(brewer.pal(8, "Set2"))(length(unique(frac_long$cell_type))))
```


# More features

## Using your own data

The simulator also has the option to create datasets with your own scRNAseq experiments.
You can create datasets in three different ways: \

* Using a expression matrix & annotation file

```{r}
# change paths accordingly
counts <- Matrix(as.matrix(fread("~/ma/data/Maynard/X_tpm.csv")), sparse = T)
genes <- fread("~/ma/data/Maynard/var.csv")
cells <- fread("~/ma/data/Maynard/obs.csv")
cellnames <- cells$Run
genenames <- genes$symbol
dimnames(counts)<-list(cellnames, genenames)
counts <- t(counts)

annotation <- fread("~/ma/data/Maynard/annotation_obs.csv")[,c("Run","cell_type")]
colnames(annotation) <- c("ID", "cell_type")

dataset_file <- simulator::dataset(annotation = annotation,
                                   count_matrix = counts,
                                   name="Maynard")
```

* Using a h5ad expression data & annotation file

```{r}
# change paths accordingly
counts <- "~/ma/data/Travaglini/Travaglini_Krasnow_2020_Lung_SS2.h5ad"
annotation <- fread("~/ma/data/Travaglini/obs_extended.csv")
# renaming the columns
colnames(annotation)[which(colnames(annotation) == "free_annotation")]<-"cell_type"
colnames(annotation)[which(colnames(annotation) == "index")]<-"ID"

dataset_h5ad <- simulator::dataset_h5ad(annotation = annotation,
                                        h5ad_file = counts, 
                                        name = "Travaglini")
```


* Using a Seurat Object and annotation file

```{r}
# download Seurat object from tabula muris
# https://figshare.com/ndownloader/files/13088531
load("~/ma/data/tabula_muris/tabula_muris_bladder.Robj")
seurat_obj <- Seurat::UpdateSeuratObject(tiss)
annotation <- seurat_obj@meta.data
# renaming the columns
colnames(annotation)[which(colnames(annotation) == "free_annotation")]<-"cell_type"
annotation$ID <- rownames(annotation)

dataset_seurat <- simulator::dataset_seurat(annotation = annotation,
                                            seurat_obj = seurat_obj,
                                            name="TabulaMuris_Bladder")
```


**Important:** the annotation dataframe you are feeding into the dataset, needs these two
columns: `ID` for the cell ID, which has to be in common with the cell IDs in the 
expression matrix and `cell_type`, which gives the cell-type for a cell. \

## Simulate using a whitelist (and blacklist) of cell-types

Sometimes, you are only interested in specific cell-types, but the dataset you are using
has way too many cell-types:

```{r}
unique(dataset_file@annotation$cell_type)
length(unique(dataset_file@annotation$cell_type))
```
You can handle this issue during simulation using the `whitelist` parameter:

```{r, fig.width=8, fig.height=8}
simulation <-  simulator::simulate_bulk(data = dataset_file,
                                        scenario = "spike-in", 
                                        scaling_factor = "none", 
                                        spike_in_cell_type = "NK cell",
                                        spike_in_amount = 0.5,
                                        ncells=1000, 
                                        nsamples = 50, 
                                        ncores = 12,
                                        whitelist = c("NK cell","B cell","T cell CD8", "T cell CD4", "Macrophage", "Monocyte conventional", "Monocyte non-conventional"))

fractions <- simulation$cell_fractions
fractions$sample <- rownames(fractions)
frac_long <- gather(fractions, cell_type, fraction, 1:length(fractions)-1)

ggplot(frac_long, aes(x=fraction, y=sample, fill=cell_type))+
  geom_col()+
  ggtitle("Cell-type fractions for 50 pseudo-bulk RNAseq samples")+
  scale_fill_manual(values = colorRampPalette(brewer.pal(8, "Set2"))(length(unique(frac_long$cell_type))))
```

In the same way, you can also provide a `blacklist` parameter, where you name the
cell-types you **don't** want to be included in your simulation.

## Using scaling factors (Census)
[I am still working on this section..]

[Census](http://cole-trapnell-lab.github.io/monocle-release/docs/#census) is an 
approach which tries to convert TPM counts into relative transcript counts. This 
basically means, you get the mRNA counts per cell, which can differ between cell-types. \
[@Qiu2017] state in their paper, that it should only be applied to TPM/FPKM normalized
data, but I tried it out with raw expression counts as well, which worked as well. \
Census calculates a vector with a scaling value for each cell in a sample. The simulated
expression matrix will be multiplied with this vector in order to census-normalize the 
pseudo-bulk sample. \
You can switch this feature on, by setting the `scaling_factor` parameter to `census`.

```{r}
simulation <-  simulator::simulate_bulk(data = dataset_file,
                                        scenario = "spike-in", 
                                        scaling_factor = "census",   # use census scaling factor
                                        spike_in_cell_type = "NK cell",
                                        spike_in_amount = 0.5,
                                        ncells=1000, 
                                        nsamples = 50, 
                                        ncores = 12,
                                        whitelist = c("NK cell","B cell","T cell CD8", "T cell CD4", "Macrophage", "Monocyte conventional", "Monocyte non-conventional"))
```



# References
