<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SimBu">
<title>Public Data Integration using Sfaira • SimBu</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Public Data Integration using Sfaira">
<meta property="og:description" content="SimBu">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SimBu</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/simulator_documentation.html">Getting started with the pseudo-bulk simulator R package</a>
    <a class="dropdown-item" href="../articles/simulator_input_output.html">Inputs &amp; Outputs</a>
    <a class="dropdown-item" href="../articles/sfaira_vignette.html">Public Data Integration using Sfaira</a>
    <a class="dropdown-item" href="../articles/simulator_scaling_factors.html">Introducing mRNA bias into simulations with scaling factors</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/omnideconv/SimBu/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Public Data Integration using Sfaira</h1>
                        <h4 data-toc-skip class="author">Alexander Dietrich</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/omnideconv/SimBu/blob/HEAD/vignettes/sfaira_vignette.Rmd" class="external-link"><code>vignettes/sfaira_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>sfaira_vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/omnideconv/SimBu" class="external-link">SimBu</a></span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="sfaira-integeration">Sfaira Integeration<a class="anchor" aria-label="anchor" href="#sfaira-integeration"></a>
</h2>
<p>This vignette will cover the integration of the public database Sfaria.<br></p>
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<p>As a public database, sfaira <span class="citation">(<a href="#ref-Fischer2020" role="doc-biblioref">Fischer et al. 2020</a>)</span> is used, which is a dataset and model repository for single-cell RNA-sequencing data. It gives access to about 233 datasets from human and mouse with more than 3 million cells in total. You can browse them interactively here: <a href="https://theislab.github.io/sfaira-portal/Datasets" class="external-link">https://theislab.github.io/sfaira-portal/Datasets</a>. Note that only annotated datasets will be downloaded!<br>
In order to use this database, we first need to install it. This can easily be done, by running the <code><a href="../reference/setup_sfaira.html">setup_sfaira()</a></code> function for the first time. In the background we use the <a href="https://bioconductor.org/packages/release/bioc/html/basilisk.utils.html" class="external-link">basilisik</a> package to establish a conda environment that has all sfaira dependencies installed. The installation will be only performed one single time, even if you close your R session and call <code><a href="../reference/setup_sfaira.html">setup_sfaira()</a></code> again. This directory serves as the storage for all future downloaded datasets from sfaira:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">setup_list</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/setup_sfaira.html">setup_sfaira</a></span><span class="op">(</span>basedir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "Sucessfully loaded sfaira."</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="creating-a-dataset">Creating a dataset<a class="anchor" aria-label="anchor" href="#creating-a-dataset"></a>
</h3>
<p>The simulator package works with a internally defined data-structure: <em>datasets</em><br>
We will now create a dataset of samples from human pancreas using the <code>organisms</code> and <code>tissues</code> parameter. You can provide a single word (like we do here) or for example a list of tissues you want to download: <code>c("pancreas","lung")</code>. An additional parameter is the <code>assays</code> parameter, where you subset the database further to only download datasets from certain sequencing assays (for examples <code>Smart-seq2</code>).<br>
The <code>name</code> parameter is used later on to give each sample (cell) a unique name.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds_pancrease</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/dataset_sfaira_multiple.html">dataset_sfaira_multiple</a></span><span class="op">(</span>sfaira_setup <span class="op">=</span> <span class="va">setup_list</span>,
                                               organisms <span class="op">=</span> <span class="st">"Homo sapiens"</span>, 
                                               tissues <span class="op">=</span> <span class="st">"pancreas"</span>, 
                                               name<span class="op">=</span><span class="st">"human_pancreas"</span><span class="op">)</span></code></pre></div>
<p>Currently there are three datasets in sfaira from human pancreas, which have cell-type annotation. The package will download them for you automatically and merge them together into a single expression matrix and a streamlined annotation table, which we can use for our simulation.<br>
It can happen, that some datasets from sfaira are not (yet) ready for the automatic download, an error message will then appear in R, telling you which file to download and where to put it.<br></p>
<p>If you wish to see all datasets which are included in sfaira you can use the following command:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">all_datasets</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/sfaira_overview.html">sfaira_overview</a></span><span class="op">(</span>setup_list <span class="op">=</span> <span class="va">setup_list</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">all_datasets</span><span class="op">)</span>
<span class="co">#&gt;                                                                                                id    author                       doi</span>
<span class="co">#&gt; 1:                            homosapiens_liver_2019_10x3v2_popescu_001_10.1038/s41586-019-1652-y   Popescu 10.1038/s41586-019-1652-y</span>
<span class="co">#&gt; 2:                 homosapiens_lungparenchyma_2019_10x3v2_madissoon_001_10.1186/s13059-019-1906-x Madissoon 10.1186/s13059-019-1906-x</span>
<span class="co">#&gt; 3:                      homosapiens_esophagus_2019_10x3v2_madissoon_002_10.1186/s13059-019-1906-x Madissoon 10.1186/s13059-019-1906-x</span>
<span class="co">#&gt; 4:                         homosapiens_spleen_2019_10x3v2_madissoon_003_10.1186/s13059-019-1906-x Madissoon 10.1186/s13059-019-1906-x</span>
<span class="co">#&gt; 5:                              homosapiens_lung_2019_dropseq_braga_001_10.1038/s41591-019-0468-5     Braga 10.1038/s41591-019-0468-5</span>
<span class="co">#&gt; 6: homosapiens_lungparenchyma_2019_10x3transcriptionprofiling_braga_001_10.1038/s41591-019-0468-5     Braga 10.1038/s41591-019-0468-5</span>
<span class="co">#&gt;    annotated                          assay           organ     organism</span>
<span class="co">#&gt; 1:      TRUE                      10x 3' v2           liver Homo sapiens</span>
<span class="co">#&gt; 2:      TRUE                      10x 3' v2 lung parenchyma Homo sapiens</span>
<span class="co">#&gt; 3:      TRUE                      10x 3' v2       esophagus Homo sapiens</span>
<span class="co">#&gt; 4:      TRUE                      10x 3' v2          spleen Homo sapiens</span>
<span class="co">#&gt; 5:      TRUE                       Drop-seq            lung Homo sapiens</span>
<span class="co">#&gt; 6:      TRUE 10x 3' transcription profiling lung parenchyma Homo sapiens</span></code></pre></div>
<p>This allows you to find the specific IDs of datasets, which you can download directly:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/dataset_sfaira.html">dataset_sfaira</a></span><span class="op">(</span>sfaira_id <span class="op">=</span> <span class="st">'homosapiens_lungparenchyma_2019_10x3v2_madissoon_001_10.1186/s13059-019-1906-x'</span>,
                      sfaira_setup <span class="op">=</span> <span class="va">setup_list</span>,
                      name <span class="op">=</span> <span class="st">"dataset_by_id"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Starting to download dataset from Sfaria with id: homosapiens_lungparenchyma_2019_10x3v2_madissoon_001_10.1186/s13059-019-1906-x"</span>
<span class="co">#&gt; [1] "Downloading datasets..."</span>
<span class="co">#&gt; [1] "Streamlining features &amp; meta-data..."</span>
<span class="co">#&gt; Using rownames for cell-IDs.</span>
<span class="co">#&gt; Filtering genes...</span>
<span class="co">#&gt; Created dataset.</span>
<span class="co">#&gt; class: SummarizedExperiment </span>
<span class="co">#&gt; dim: 20195 57020 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames(20195): TSPAN6 TNMD ... LINC02498 MGC4859</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames(57020): dataset_by_id_1 dataset_by_id_2 ... dataset_by_id_57019 dataset_by_id_57020</span>
<span class="co">#&gt; colData names(5): cell_ID cell_ID.old cell_type nReads_SimBu nGenes_SimBu</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.0 (2021-05-18)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.6 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] SimBu_0.99.0                SummarizedExperiment_1.22.0 Biobase_2.52.0              GenomicRanges_1.44.0       </span>
<span class="co">#&gt;  [5] GenomeInfoDb_1.28.4         IRanges_2.26.0              S4Vectors_0.30.2            BiocGenerics_0.38.0        </span>
<span class="co">#&gt;  [9] MatrixGenerics_1.4.3        matrixStats_0.61.0          Matrix_1.3-4               </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] plyr_1.8.7              igraph_1.2.11           lazyeval_0.2.2          proxyC_0.2.4            splines_4.1.0          </span>
<span class="co">#&gt;   [6] listenv_0.8.0           scattermore_0.8         ggplot2_3.3.5           digest_0.6.27           htmltools_0.5.2        </span>
<span class="co">#&gt;  [11] fansi_1.0.3             magrittr_2.0.1          memoise_2.0.1           tensor_1.5              cluster_2.1.2          </span>
<span class="co">#&gt;  [16] ROCR_1.0-11             globals_0.14.0          RcppParallel_5.1.5      pkgdown_2.0.2           spatstat.sparse_2.1-0  </span>
<span class="co">#&gt;  [21] colorspace_2.0-3        ggrepel_0.9.1           xfun_0.30               dplyr_1.0.8             callr_3.7.0            </span>
<span class="co">#&gt;  [26] crayon_1.5.1            RCurl_1.98-1.6          jsonlite_1.7.2          graph_1.70.0            spatstat.data_2.1-2    </span>
<span class="co">#&gt;  [31] survival_3.2-11         zoo_1.8-9               glue_1.6.2              polyclip_1.10-0         gtable_0.3.0           </span>
<span class="co">#&gt;  [36] zlibbioc_1.38.0         XVector_0.32.0          leiden_0.3.9            DelayedArray_0.18.0     future.apply_1.8.1     </span>
<span class="co">#&gt;  [41] abind_1.4-5             scales_1.1.1            DBI_1.1.2               spatstat.random_2.1-0   miniUI_0.1.1.1         </span>
<span class="co">#&gt;  [46] Rcpp_1.0.8.3            viridisLite_0.4.0       xtable_1.8-4            reticulate_1.24         spatstat.core_2.4-0    </span>
<span class="co">#&gt;  [51] getopt_1.20.3           htmlwidgets_1.5.4       httr_1.4.2              dir.expiry_1.0.0        RColorBrewer_1.1-2     </span>
<span class="co">#&gt;  [56] ellipsis_0.3.2          Seurat_4.1.0            ica_1.0-2               XML_4.0-0               pkgconfig_2.0.3        </span>
<span class="co">#&gt;  [61] uwot_0.1.11             deldir_1.0-6            here_1.0.1              utf8_1.2.2              tidyselect_1.1.2       </span>
<span class="co">#&gt;  [66] rlang_1.0.2             reshape2_1.4.4          later_1.3.0             biocViews_1.60.0        munsell_0.5.0          </span>
<span class="co">#&gt;  [71] tools_4.1.0             cachem_1.0.6            cli_3.2.0               generics_0.1.2          ggridges_0.5.3         </span>
<span class="co">#&gt;  [76] evaluate_0.14           stringr_1.4.0           fastmap_1.1.0           yaml_2.3.5              goftest_1.2-3          </span>
<span class="co">#&gt;  [81] processx_3.5.3          knitr_1.33              fs_1.5.2                fitdistrplus_1.1-8      purrr_0.3.4            </span>
<span class="co">#&gt;  [86] RANN_2.6.1              sparseMatrixStats_1.4.2 RBGL_1.68.0             pbapply_1.5-0           future_1.24.0          </span>
<span class="co">#&gt;  [91] nlme_3.1-152            mime_0.10               xml2_1.3.3              rstudioapi_0.13         brio_1.1.3             </span>
<span class="co">#&gt;  [96] compiler_4.1.0          curl_4.3.2              plotly_4.10.0           filelock_1.0.2          png_0.1-7              </span>
<span class="co">#&gt; [101] testthat_3.1.2          spatstat.utils_2.3-0    tibble_3.1.6            stringi_1.6.1           ps_1.6.0               </span>
<span class="co">#&gt; [106] basilisk.utils_1.4.0    desc_1.4.1              lattice_0.20-44         commonmark_1.8.0        vctrs_0.3.8            </span>
<span class="co">#&gt; [111] stringdist_0.9.8        BiocCheck_1.28.0        pillar_1.7.0            lifecycle_1.0.1         RUnit_0.4.32           </span>
<span class="co">#&gt; [116] BiocManager_1.30.16     optparse_1.7.1          spatstat.geom_2.3-2     lmtest_0.9-40           RcppAnnoy_0.0.19       </span>
<span class="co">#&gt; [121] data.table_1.14.2       cowplot_1.1.1           bitops_1.0-7            irlba_2.3.5             httpuv_1.6.5           </span>
<span class="co">#&gt; [126] patchwork_1.1.1         R6_2.5.1                promises_1.2.0.1        KernSmooth_2.23-20      gridExtra_2.3          </span>
<span class="co">#&gt; [131] parallelly_1.30.0       codetools_0.2-18        MASS_7.3-54             assertthat_0.2.1        rprojroot_2.0.2        </span>
<span class="co">#&gt; [136] withr_2.5.0             SeuratObject_4.0.4      sctransform_0.3.3       GenomeInfoDbData_1.2.6  mgcv_1.8-36            </span>
<span class="co">#&gt; [141] grid_4.1.0              rpart_4.1-15            tidyr_1.2.0             basilisk_1.4.0          rmarkdown_2.13         </span>
<span class="co">#&gt; [146] downlit_0.4.0           Rtsne_0.15              shiny_1.7.1</span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Fischer2020" class="csl-entry">
Fischer, David S., Leander Dony, Martin König, Abdul Moeed, Luke Zappia, Sophie Tritschler, Olle Holmberg, Hananeh Aliee, and Fabian J. Theis. 2020. <span>“Sfaira Accelerates Data and Model Reuse in Single Cell Genomics.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2020.12.16.419036" class="external-link">https://doi.org/10.1101/2020.12.16.419036</a>.
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Alexander Dietrich.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
