<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SimBu">
<title>Introducing mRNA bias into simulations with scaling factors • SimBu</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introducing mRNA bias into simulations with scaling factors">
<meta property="og:description" content="SimBu">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SimBu</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/simulator_documentation.html">Getting started with the pseudo-bulk simulator R package</a>
    <a class="dropdown-item" href="../articles/simulator_input_output.html">Inputs &amp; Outputs</a>
    <a class="dropdown-item" href="../articles/sfaira_vignette.html">Public Data Integration using Sfaira</a>
    <a class="dropdown-item" href="../articles/simulator_scaling_factors.html">Introducing mRNA bias into simulations with scaling factors</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/omnideconv/SimBu/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introducing mRNA bias into simulations with scaling factors</h1>
                        <h4 data-toc-skip class="author">Alexander Dietrich</h4>
            
            <h4 data-toc-skip class="date">11/29/2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/omnideconv/SimBu/blob/HEAD/vignettes/simulator_scaling_factors.Rmd" class="external-link"><code>vignettes/simulator_scaling_factors.Rmd</code></a></small>
      <div class="d-none name"><code>simulator_scaling_factors.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="using-scaling-factors">Using scaling factors<a class="anchor" aria-label="anchor" href="#using-scaling-factors"></a>
</h2>
<p>This package allows the user to introduce an mRNA bias into pseudo-bulk RNA-seq samples. Different cell-types contain different amounts of mRNA (Dendritic cells for examples contain much more than Neutrophils); this bias can be added into simulations artificially in different ways.<br>
The scaling factors will always be applied on the single-cell dataset first, altering its expression profiles accordingly, and then the pseudo-bulk samples are generated by summing up the count data from the sampled cells.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Example data</span>
<span class="va">counts</span> <span class="op">&lt;-</span>  <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html" class="external-link">Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">3e5</span>, <span class="fl">5</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">300</span><span class="op">)</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">tpm</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html" class="external-link">Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">3e5</span>, <span class="fl">5</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">300</span><span class="op">)</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">tpm</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="fl">1e6</span><span class="op">*</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span><span class="op">/</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">300</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">300</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span>  

<span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"ID"</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">300</span><span class="op">)</span><span class="op">)</span>,
                           <span class="st">"cell_type"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"T cells CD4"</span>,<span class="fl">150</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"T cells CD8"</span>,<span class="fl">150</span><span class="op">)</span><span class="op">)</span>,
                           <span class="st">'spikes'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">300</span><span class="op">)</span>,
                           <span class="st">'add_1'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">300</span><span class="op">)</span>,
                           <span class="st">'add_2'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">300</span><span class="op">)</span><span class="op">)</span>

<span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/dataset.html">dataset</a></span><span class="op">(</span>annotation <span class="op">=</span> <span class="va">annotation</span>,
                     count_matrix <span class="op">=</span> <span class="va">counts</span>,
                     name <span class="op">=</span> <span class="st">"test_dataset"</span><span class="op">)</span>
<span class="co">#&gt; Filtering genes...</span>
<span class="co">#&gt; Created dataset.</span></code></pre></div>
<div class="section level3">
<h3 id="pre-defined-scaling-factors">Pre-defined scaling factors<a class="anchor" aria-label="anchor" href="#pre-defined-scaling-factors"></a>
</h3>
<p>Some studies have proposed scaling factors for immune cells, such as EPIC <span class="citation">(<a href="#ref-Racle2017" role="doc-biblioref">Racle et al. 2017</a>)</span> or quanTIseq <span class="citation">(<a href="#ref-Finotello2019" role="doc-biblioref">Finotello et al. 2019</a>)</span>, deconvolution tools which correct for the mRNA bias internally using these values:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">epic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>
  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cells"</span>, 
           <span class="st">"Macrophages"</span>,
           <span class="st">"Monocytes"</span>, 
           <span class="st">"Neutrophils"</span>,
           <span class="st">"NK cells"</span>,
           <span class="st">"T cells"</span>,
           <span class="st">"T cells CD4"</span>,
           <span class="st">"T cells CD8"</span>,
           <span class="st">"T helper cells"</span>,
           <span class="st">"T regulatory cells"</span>,
           <span class="st">"otherCells"</span>,
           <span class="st">"default"</span><span class="op">)</span>,
  mRNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4016</span>,
           <span class="fl">1.4196</span>,
           <span class="fl">1.4196</span>,
           <span class="fl">0.1300</span>,
           <span class="fl">0.4396</span>,
           <span class="fl">0.3952</span>,
           <span class="fl">0.3952</span>,
           <span class="fl">0.3952</span>,
           <span class="fl">0.3952</span>,
           <span class="fl">0.3952</span>,
           <span class="fl">0.4000</span>,
           <span class="fl">0.4000</span><span class="op">)</span>
  
<span class="op">)</span>

<span class="va">epic</span>
<span class="co">#&gt;                  type   mRNA</span>
<span class="co">#&gt; 1             B cells 0.4016</span>
<span class="co">#&gt; 2         Macrophages 1.4196</span>
<span class="co">#&gt; 3           Monocytes 1.4196</span>
<span class="co">#&gt; 4         Neutrophils 0.1300</span>
<span class="co">#&gt; 5            NK cells 0.4396</span>
<span class="co">#&gt; 6             T cells 0.3952</span>
<span class="co">#&gt; 7         T cells CD4 0.3952</span>
<span class="co">#&gt; 8         T cells CD8 0.3952</span>
<span class="co">#&gt; 9      T helper cells 0.3952</span>
<span class="co">#&gt; 10 T regulatory cells 0.3952</span>
<span class="co">#&gt; 11         otherCells 0.4000</span>
<span class="co">#&gt; 12            default 0.4000</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quantiseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>
  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cells"</span>,
           <span class="st">"Macrophages"</span>,
           <span class="st">"MacrophagesM2"</span>,
           <span class="st">"Monocytes"</span>,
           <span class="st">"Neutrophils"</span>,
           <span class="st">"NK cells"</span>,
           <span class="st">"T cells CD4"</span>,
           <span class="st">"T cells CD8"</span>,
           <span class="st">"T regulatory cells"</span>,
           <span class="st">"Dendritic cells"</span>,
           <span class="st">"T cells"</span><span class="op">)</span>,
  mRNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="fl">65.66148</span>,
            <span class="fl">138.11520</span>,
            <span class="fl">119.35447</span>,
            <span class="fl">130.65455</span>,
            <span class="fl">27.73634</span>,
            <span class="fl">117.71584</span>,
            <span class="fl">63.87200</span>,
            <span class="fl">70.25659</span>,
            <span class="fl">72.55110</span>,
            <span class="fl">140.76091</span>,
            <span class="fl">68.89323</span><span class="op">)</span>
  
<span class="op">)</span>

<span class="va">quantiseq</span>
<span class="co">#&gt;                  type      mRNA</span>
<span class="co">#&gt; 1             B cells  65.66148</span>
<span class="co">#&gt; 2         Macrophages 138.11520</span>
<span class="co">#&gt; 3       MacrophagesM2 119.35447</span>
<span class="co">#&gt; 4           Monocytes 130.65455</span>
<span class="co">#&gt; 5         Neutrophils  27.73634</span>
<span class="co">#&gt; 6            NK cells 117.71584</span>
<span class="co">#&gt; 7         T cells CD4  63.87200</span>
<span class="co">#&gt; 8         T cells CD8  70.25659</span>
<span class="co">#&gt; 9  T regulatory cells  72.55110</span>
<span class="co">#&gt; 10    Dendritic cells 140.76091</span>
<span class="co">#&gt; 11            T cells  68.89323</span></code></pre></div>
<p>When you want to apply one of these scaling factors into your simulation (therefore in-/decreasing the expression signals for the cell-types), we can use the <code>scaling_factor</code> parameter. Note, that these pre-defined scaling factors only offer values for a certain number of cell types, and your annotation in the provided dataset has to match these names 1:1. All cell types from your dataset which are not present in this scaling factor remain unscaled and a warning message will appear.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sim_epic</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_bulk.html">simulate_bulk</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">ds</span>,
                                 scenario <span class="op">=</span> <span class="st">"random"</span>,
                                 scaling_factor<span class="op">=</span><span class="st">"epic"</span>,
                                 nsamples <span class="op">=</span> <span class="fl">10</span>, 
                                 ncells<span class="op">=</span><span class="fl">100</span>,
                                 BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,  
                                 run_parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Using parallel generation of simulations.</span>
<span class="co">#&gt; Using EPIC scaling factors.</span>
<span class="co">#&gt; Finished simulation.</span></code></pre></div>
<p>We can also try out some custom scaling factors, for example increasing the expression levels for a single cell-type (T cells CD8) by 10-fold compared to the rest. All cell-types which are not mentioned in the named list given to <code>custom_scaling_vector</code> will be transformed with a scaling factor of 1, meaning nothing changes for them.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sim_extreme_b</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_bulk.html">simulate_bulk</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ds</span>, 
                                      scenario <span class="op">=</span> <span class="st">"random"</span>,
                                      scaling_factor <span class="op">=</span> <span class="st">"custom"</span>,
                                      custom_scaling_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T cells CD8"</span><span class="op">=</span><span class="fl">10</span><span class="op">)</span>, 
                                      nsamples <span class="op">=</span> <span class="fl">10</span>, 
                                      ncells <span class="op">=</span> <span class="fl">100</span>, 
                                      BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,  
                                      run_parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Using parallel generation of simulations.</span>
<span class="co">#&gt; Using custom scaling factors.</span>
<span class="co">#&gt; Warning in merge_scaling_factor(data = data, scaling_factor_values = custom_scaling_vector, : For some cell type(s) in the dataset, no</span>
<span class="co">#&gt; scaling factor is available when using custom: T cells CD4. This cell type will not be re-scaled.</span>
<span class="co">#&gt; Finished simulation.</span></code></pre></div>
<p><strong>Important:</strong> Watch out that the cell-type annotation names in your dataset are the same as in the scaling factor! Otherwise the scaling factor will not be applied or even worse, applied to a different cell-type.</p>
</div>
<div class="section level3">
<h3 id="dataset-specific-scaling-factors">Dataset specific scaling factors<a class="anchor" aria-label="anchor" href="#dataset-specific-scaling-factors"></a>
</h3>
<p>You can also choose to calculate scaling factors, which are depending on your provided single-cell dataset. Compared to the previous section, this will give a unique value for each cell rather than a cell-type, making it possibly more sensitive.</p>
<div class="section level4">
<h4 id="reads-and-genes">Reads and genes<a class="anchor" aria-label="anchor" href="#reads-and-genes"></a>
</h4>
<p>Two straight forward approaches would be the number of reads or number of expressed genes/features. As these values are easily obtainable from the provided count data, SimBu already calculates them during dataset generation.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">sim_reads</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_bulk.html">simulate_bulk</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">ds</span>, 
                           scenario <span class="op">=</span> <span class="st">"random"</span>, 
                           scaling_factor<span class="op">=</span><span class="st">"read_number"</span>,  <span class="co"># use number of reads as scaling factor</span>
                           nsamples <span class="op">=</span> <span class="fl">10</span>, 
                           ncells<span class="op">=</span><span class="fl">100</span>, 
                           BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,  
                           run_parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Using parallel generation of simulations.</span>
<span class="co">#&gt; Finished simulation.</span>

<span class="va">sim_genes</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_bulk.html">simulate_bulk</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ds</span>,
                           scenario <span class="op">=</span> <span class="st">"random"</span>,
                           scaling_factor <span class="op">=</span> <span class="st">"expressed_genes"</span>, <span class="co"># use number of expressed genes column as scaling factor</span>
                           nsamples <span class="op">=</span> <span class="fl">10</span>,
                           ncells <span class="op">=</span> <span class="fl">100</span>,
                           BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,  
                           run_parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Using parallel generation of simulations.</span>
<span class="co">#&gt; Finished simulation.</span></code></pre></div>
<p>These options would also allow you to use other numerical measurements you have for the single cells as scaling factors, such as weight or size for example. Lets pretend, <code>add_1</code> and <code>add_2</code> are such measurements. With the <code>additional_cols</code> parameter, they can be added to the SimBu dataset and we can use them as scaling factor as well:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span>
<span class="co">#&gt;       ID   cell_type     spikes     add_1       add_2</span>
<span class="co">#&gt; 1 cell_1 T cells CD4 0.09136481 0.4811858 0.435582586</span>
<span class="co">#&gt; 2 cell_2 T cells CD4 0.61761242 0.6596912 0.358060187</span>
<span class="co">#&gt; 3 cell_3 T cells CD4 0.28130156 0.1950135 0.350385112</span>
<span class="co">#&gt; 4 cell_4 T cells CD4 0.44271955 0.6874199 0.008552276</span>
<span class="co">#&gt; 5 cell_5 T cells CD4 0.69949660 0.2325767 0.595156309</span>
<span class="co">#&gt; 6 cell_6 T cells CD4 0.88835517 0.3133121 0.328329093</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/dataset.html">dataset</a></span><span class="op">(</span>annotation <span class="op">=</span> <span class="va">annotation</span>,
                     count_matrix <span class="op">=</span> <span class="va">counts</span>,
                     name <span class="op">=</span> <span class="st">"test_dataset"</span>,
                     additional_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'add_1'</span>,<span class="st">'add_2'</span><span class="op">)</span><span class="op">)</span> <span class="co"># add columns to dataset</span>
<span class="co">#&gt; Filtering genes...</span>
<span class="co">#&gt; Created dataset.</span>

<span class="va">sim_genes</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_bulk.html">simulate_bulk</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ds</span>,
                           scenario <span class="op">=</span> <span class="st">"random"</span>,
                           scaling_factor <span class="op">=</span> <span class="st">"add_1"</span>, <span class="co"># use add_1 column as scaling factor</span>
                           nsamples <span class="op">=</span> <span class="fl">10</span>,
                           ncells <span class="op">=</span> <span class="fl">100</span>,
                           BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,  
                           run_parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Using parallel generation of simulations.</span>
<span class="co">#&gt; Scaling by add_1-column in annotation table; if no scaling is wished instead, use 'NONE'.</span>
<span class="co">#&gt; Finished simulation.</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="spike-ins">Spike-ins<a class="anchor" aria-label="anchor" href="#spike-ins"></a>
</h4>
<p>One other numerical measurement can be spike-ins. Usually the number of reads mapped to spike-in molecules per cell is given in the cell annotations. If this is the case, they can be stored in the dataset annotation using the <code>spike_in_col</code> parameter, where you indicate the name of the column from the <code>annotation</code> dataframe in which the spike-in information is stored. To calculate a scaling factor from this, the number of reads are also necessary, so we will add this information as well (as above using the <code>read_number_col</code> parameter).</p>
<p>The scaling factor with spike-ins is calculated as the “% of reads NOT mapped to spike-in reads,” or: <code>(n_reads - n_spike_in)/n_reads</code> for each cell. We apply it like this:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/dataset.html">dataset</a></span><span class="op">(</span>annotation <span class="op">=</span> <span class="va">annotation</span>,
                     count_matrix <span class="op">=</span> <span class="va">counts</span>,
                     name <span class="op">=</span> <span class="st">"test_dataset"</span>,
                     spike_in_col <span class="op">=</span> <span class="st">'spikes'</span><span class="op">)</span>    <span class="co"># give the name in the annotation file, that contains spike-in information</span>

<span class="va">sim_spike</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_bulk.html">simulate_bulk</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ds</span>,
                           scenario <span class="op">=</span> <span class="st">"random"</span>,
                           scaling_factor <span class="op">=</span> <span class="st">"spike_in"</span>, <span class="co"># use spike-in scaling factor</span>
                           nsamples <span class="op">=</span> <span class="fl">10</span>,
                           ncells <span class="op">=</span> <span class="fl">100</span>,
                           BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,  
                           run_parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="census---estimate-mrna-counts-per-cell">Census - estimate mRNA counts per cell<a class="anchor" aria-label="anchor" href="#census---estimate-mrna-counts-per-cell"></a>
</h3>
<p><a href="http://cole-trapnell-lab.github.io/monocle-release/docs/#census" class="external-link">Census</a> is an approach which tries to convert TPM counts into relative transcript counts. This basically means, you get the mRNA counts per cell, which can differ between cell-types.<br><span class="citation">(<a href="#ref-Qiu2017" role="doc-biblioref">Qiu et al. 2017</a>)</span> state in their paper, that it should only be applied to TPM/FPKM normalized data, but I tried it out with raw expression counts as well, which worked as well.<br>
Census calculates a vector with a scaling value for each cell in a sample. You can switch this feature on, by setting the <code>scaling_factor</code> parameter to <code>census</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/dataset.html">dataset</a></span><span class="op">(</span>annotation <span class="op">=</span> <span class="va">annotation</span>,
                     count_matrix <span class="op">=</span> <span class="va">counts</span>,
                     tpm_matrix <span class="op">=</span> <span class="va">tpm</span>,
                     name <span class="op">=</span> <span class="st">"test_dataset"</span><span class="op">)</span>
<span class="co">#&gt; Filtering genes...</span>
<span class="co">#&gt; Created dataset.</span>

<span class="va">sim_census</span> <span class="op">&lt;-</span>  <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_bulk.html">simulate_bulk</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ds</span>,
                             scenario <span class="op">=</span> <span class="st">"random"</span>, 
                             scaling_factor <span class="op">=</span> <span class="st">"census"</span>,   <span class="co"># use census scaling factor</span>
                             nsamples <span class="op">=</span> <span class="fl">10</span>, 
                             ncells<span class="op">=</span><span class="fl">100</span>, 
                             BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,  
                             run_parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Using parallel generation of simulations.</span>
<span class="co">#&gt; Finished simulation.</span></code></pre></div>
<p>In our analysis we found, that Census is basically a complicated way of estimating the number of expressed genes per cell. It will remain to the user to decide if he/she wants to use census or simply the number of expressed genes (as shown above) as scaling factor.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.0 (2021-05-18)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.6 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] curl_4.3.2                  SeuratObject_4.0.4          Seurat_4.1.0                SimBu_0.99.0               </span>
<span class="co">#&gt;  [5] SummarizedExperiment_1.22.0 Biobase_2.52.0              GenomicRanges_1.44.0        GenomeInfoDb_1.28.4        </span>
<span class="co">#&gt;  [9] IRanges_2.26.0              S4Vectors_0.30.2            BiocGenerics_0.38.0         MatrixGenerics_1.4.3       </span>
<span class="co">#&gt; [13] matrixStats_0.61.0          Matrix_1.3-4               </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] utf8_1.2.2              reticulate_1.24         RUnit_0.4.32            tidyselect_1.1.2        htmlwidgets_1.5.4      </span>
<span class="co">#&gt;   [6] grid_4.1.0              BiocParallel_1.26.2     Rtsne_0.15              munsell_0.5.0           codetools_0.2-18       </span>
<span class="co">#&gt;  [11] ica_1.0-2               future_1.24.0           miniUI_0.1.1.1          withr_2.5.0             spatstat.random_2.1-0  </span>
<span class="co">#&gt;  [16] colorspace_2.0-3        filelock_1.0.2          highr_0.9               knitr_1.33              rstudioapi_0.13        </span>
<span class="co">#&gt;  [21] ROCR_1.0-11             tensor_1.5              listenv_0.8.0           labeling_0.4.2          optparse_1.7.1         </span>
<span class="co">#&gt;  [26] GenomeInfoDbData_1.2.6  polyclip_1.10-0         farver_2.1.0            rprojroot_2.0.2         basilisk_1.4.0         </span>
<span class="co">#&gt;  [31] parallelly_1.30.0       vctrs_0.3.8             generics_0.1.2          xfun_0.30               R6_2.5.1               </span>
<span class="co">#&gt;  [36] bitops_1.0-7            spatstat.utils_2.3-0    cachem_1.0.6            DelayedArray_0.18.0     assertthat_0.2.1       </span>
<span class="co">#&gt;  [41] promises_1.2.0.1        scales_1.1.1            gtable_0.3.0            downlit_0.4.0           biocViews_1.60.0       </span>
<span class="co">#&gt;  [46] globals_0.14.0          processx_3.5.3          goftest_1.2-3           rlang_1.0.2             splines_4.1.0          </span>
<span class="co">#&gt;  [51] lazyeval_0.2.2          spatstat.geom_2.3-2     BiocManager_1.30.16     yaml_2.3.5              reshape2_1.4.4         </span>
<span class="co">#&gt;  [56] abind_1.4-5             httpuv_1.6.5            RBGL_1.68.0             tools_4.1.0             ggplot2_3.3.5          </span>
<span class="co">#&gt;  [61] ellipsis_0.3.2          spatstat.core_2.4-0     RColorBrewer_1.1-2      ggridges_0.5.3          Rcpp_1.0.8.3           </span>
<span class="co">#&gt;  [66] plyr_1.8.7              sparseMatrixStats_1.4.2 zlibbioc_1.38.0         purrr_0.3.4             RCurl_1.98-1.6         </span>
<span class="co">#&gt;  [71] basilisk.utils_1.4.0    ps_1.6.0                rpart_4.1-15            deldir_1.0-6            pbapply_1.5-0          </span>
<span class="co">#&gt;  [76] cowplot_1.1.1           zoo_1.8-9               ggrepel_0.9.1           cluster_2.1.2           fs_1.5.2               </span>
<span class="co">#&gt;  [81] here_1.0.1              magrittr_2.0.1          data.table_1.14.2       scattermore_0.8         lmtest_0.9-40          </span>
<span class="co">#&gt;  [86] RANN_2.6.1              fitdistrplus_1.1-8      patchwork_1.1.1         mime_0.10               evaluate_0.14          </span>
<span class="co">#&gt;  [91] xtable_1.8-4            XML_4.0-0               gridExtra_2.3           testthat_3.1.2          compiler_4.1.0         </span>
<span class="co">#&gt;  [96] tibble_3.1.6            KernSmooth_2.23-20      crayon_1.5.1            htmltools_0.5.2         proxyC_0.2.4           </span>
<span class="co">#&gt; [101] mgcv_1.8-36             later_1.3.0             tidyr_1.2.0             RcppParallel_5.1.5      DBI_1.1.2              </span>
<span class="co">#&gt; [106] MASS_7.3-54             getopt_1.20.3           brio_1.1.3              cli_3.2.0               igraph_1.2.11          </span>
<span class="co">#&gt; [111] pkgconfig_2.0.3         pkgdown_2.0.2           dir.expiry_1.0.0        plotly_4.10.0           spatstat.sparse_2.1-0  </span>
<span class="co">#&gt; [116] xml2_1.3.3              stringdist_0.9.8        XVector_0.32.0          BiocCheck_1.28.0        stringr_1.4.0          </span>
<span class="co">#&gt; [121] callr_3.7.0             digest_0.6.27           sctransform_0.3.3       RcppAnnoy_0.0.19        graph_1.70.0           </span>
<span class="co">#&gt; [126] spatstat.data_2.1-2     rmarkdown_2.13          leiden_0.3.9            uwot_0.1.11             shiny_1.7.1            </span>
<span class="co">#&gt; [131] commonmark_1.8.0        lifecycle_1.0.1         nlme_3.1-152            jsonlite_1.7.2          desc_1.4.1             </span>
<span class="co">#&gt; [136] viridisLite_0.4.0       fansi_1.0.3             pillar_1.7.0            lattice_0.20-44         fastmap_1.1.0          </span>
<span class="co">#&gt; [141] httr_1.4.2              survival_3.2-11         glue_1.6.2              png_0.1-7               stringi_1.6.1          </span>
<span class="co">#&gt; [146] memoise_2.0.1           dplyr_1.0.8             irlba_2.3.5             future.apply_1.8.1</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Finotello2019" class="csl-entry">
Finotello, Francesca, Clemens Mayer, Christina Plattner, Gerhard Laschober, Dietmar Rieder, Hubert Hackl, Anne Krogsdam, et al. 2019. <span>“Molecular and Pharmacological Modulators of the Tumor Immune Contexture Revealed by Deconvolution of RNA-Seq Data.”</span> <em>Genome Medicine</em> 11 (1): 34. <a href="https://doi.org/10.1186/s13073-019-0638-6" class="external-link">https://doi.org/10.1186/s13073-019-0638-6</a>.
</div>
<div id="ref-Qiu2017" class="csl-entry">
Qiu, Xiaojie, Andrew Hill, Jonathan Packer, Dejun Lin, Yi-An Ma, and Cole Trapnell. 2017. <span>“Single-Cell mRNA Quantification and Differential Analysis with Census.”</span> <em>Nature Methods</em> 14 (3): 309–15. <a href="https://doi.org/10.1038/nmeth.4150" class="external-link">https://doi.org/10.1038/nmeth.4150</a>.
</div>
<div id="ref-Racle2017" class="csl-entry">
Racle, Julien, Kaat de Jonge, Petra Baumgaertner, Daniel E. Speiser, and David Gfeller. 2017. <span>“Simultaneous Enumeration of Cancer and Immune Cell Types from Bulk Tumor Gene Expression Data.”</span> <em>eLife</em> 6 (November): e26476. <a href="https://doi.org/10.7554/eLife.26476" class="external-link">https://doi.org/10.7554/eLife.26476</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Alexander Dietrich.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
