<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SimBu">
<title>Inputs &amp; Outputs • SimBu</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Inputs &amp; Outputs">
<meta property="og:description" content="SimBu">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SimBu</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/simulator_documentation.html">Getting started with the pseudo-bulk simulator R package</a>
    <a class="dropdown-item" href="../articles/simulator_input_output.html">Inputs &amp; Outputs</a>
    <a class="dropdown-item" href="../articles/sfaira_vignette.html">Public Data Integration using Sfaira</a>
    <a class="dropdown-item" href="../articles/simulator_scaling_factors.html">Introducing mRNA bias into simulations with scaling factors</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/omnideconv/SimBu/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Inputs &amp; Outputs</h1>
                        <h4 data-toc-skip class="author">Alexander Dietrich</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/omnideconv/SimBu/blob/HEAD/vignettes/simulator_input_output.Rmd" class="external-link"><code>vignettes/simulator_input_output.Rmd</code></a></small>
      <div class="d-none name"><code>simulator_input_output.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="co">#&gt; Attaching SeuratObject</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'Seurat'</span>
<span class="co">#&gt; The following object is masked from 'package:SummarizedExperiment':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Assays</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jeroen/curl" class="external-link">curl</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'curl' was built under R version 4.1.2</span>
<span class="co">#&gt; Using libcurl 7.58.0 with GnuTLS/3.5.18</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/omnideconv/SimBu" class="external-link">SimBu</a></span><span class="op">)</span></code></pre></div>
<p>This chapter covers the different input and output options of the package in detail.</p>
<div class="section level2">
<h2 id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a>
</h2>
<p>The input for your simulations is always a <code>SummarizedExperiment</code> object. You can create this object with different constructing functions, which will explained below. It is also possible to merge multiple datasets objects into one.<br>
Sfaira is not covered in this vignette, but in <a href="sfaira_vignette.html">“Public Data Integration”</a>.<br></p>
<div class="section level3">
<h3 id="custom-data">Custom data<a class="anchor" aria-label="anchor" href="#custom-data"></a>
</h3>
<p>Using existing count matrices and annotations is already covered in the <a href="simulator_documentation.html">“Getting started”</a> vignette; this section will explain some minor details.<br></p>
<p>When generating a dataset with your own data, you need to provide the <code>count_matrix</code> parameter of <code><a href="../reference/dataset.html">dataset()</a></code>; additionally you can provide a TPM matrix with the <code>tpm_matrix</code>. This will then lead to two simulations, one based on counts and one based on TPMs. For either of them, genes are located in the rows, cells in the columns.<br>
Additionally, an annotation table is needed, with the cell-type annotations. It needs to consist of at least out of 2 columns: <code>ID</code> and <code>cell_type</code>, where <code>ID</code> has to be identical to the column names of the provides matrix/matrices. If not all cells appear in the annotation or matrix, the intersection of both is used to generate the dataset.<br></p>
<p>Here is some example data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html" class="external-link">Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">3e5</span>, <span class="fl">5</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">300</span><span class="op">)</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">tpm</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html" class="external-link">Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">3e5</span>, <span class="fl">5</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">300</span><span class="op">)</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">tpm</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="fl">1e6</span><span class="op">*</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span><span class="op">/</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell-"</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">300</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell-"</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">300</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene-"</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene-"</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span>

<span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"ID"</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell-"</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">300</span><span class="op">)</span><span class="op">)</span>, 
                         <span class="st">"cell_type"</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"T cells CD4"</span>,<span class="fl">50</span><span class="op">)</span>, 
                                       <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"T cells CD8"</span>,<span class="fl">50</span><span class="op">)</span>,
                                       <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Macrophages"</span>,<span class="fl">100</span><span class="op">)</span>,
                                       <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"NK cells"</span>,<span class="fl">10</span><span class="op">)</span>,
                                       <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"B cells"</span>,<span class="fl">70</span><span class="op">)</span>,
                                       <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Monocytes"</span>,<span class="fl">20</span><span class="op">)</span><span class="op">)</span>,
                         row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell-"</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">300</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>, assay <span class="op">=</span> <span class="st">'counts'</span>, meta.data <span class="op">=</span> <span class="va">annotation</span><span class="op">)</span>
<span class="va">tpm_assay</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateAssayObject.html" class="external-link">CreateAssayObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">tpm</span><span class="op">)</span>
<span class="va">seurat_obj</span><span class="op">[[</span><span class="st">'tpm'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">tpm_assay</span>

<span class="va">seurat_obj</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 2000 features across 300 samples within 2 assays </span>
<span class="co">#&gt; Active assay: counts (1000 features, 0 variable features)</span>
<span class="co">#&gt;  1 other assay present: tpm</span></code></pre></div>
<div class="section level4">
<h4 id="seurat">Seurat<a class="anchor" aria-label="anchor" href="#seurat"></a>
</h4>
<p>It is possible to use a Seurat object to build a dataset; give the name of the assay containing count data in the <code>counts</code> slot, the name of the column in the meta table with the unique cell IDs and the name of the column in the meta table with the cell type identifier. Additionally you may give the name of the assay containing TPM data in the <code>counts</code> slot.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds_seurat</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/dataset_seurat.html">dataset_seurat</a></span><span class="op">(</span>seurat_obj <span class="op">=</span> <span class="va">seurat_obj</span>, 
                                   count_assay <span class="op">=</span> <span class="st">"counts"</span>, 
                                   cell_id_col <span class="op">=</span> <span class="st">'ID'</span>, 
                                   cell_type_col <span class="op">=</span> <span class="st">'cell_type'</span>, 
                                   tpm_assay <span class="op">=</span> <span class="st">'tpm'</span>,
                                   name <span class="op">=</span> <span class="st">"seurat_dataset"</span><span class="op">)</span>
<span class="co">#&gt; Filtering genes...</span>
<span class="co">#&gt; Created dataset.</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="h5ad-files">h5ad files<a class="anchor" aria-label="anchor" href="#h5ad-files"></a>
</h4>
<p>It is possible to use an h5ad file directly, a file format which stores <a href="https://anndata.readthedocs.io/en/latest/" class="external-link">AnnData</a> objects. As h5ad files can store cell specific information in the <code>obs</code> layer, no additional annotation input for SimBu is needed.<br>
Note: if you want both counts and tpm data as input, you will have to provide two files; the cell annotation has to match between these two files. As SimBu expects the cells to be in the columns and genes/features in the rows of the input matrix, but this is not necessarily the case for <a href="https://falexwolf.de/img/scanpy/anndata.svg" class="external-link">anndata objects</a>, SimBu can handle h5ad files with cells in the <code>obs</code> or <code>var</code> layer. If your cells are in <code>obs</code>, use <code>cells_in_obs=TRUE</code> and <code>FALSE</code> otherwise. This will also automatically transpose the matrix. To know, which columns in the cell annotation layer correspond to the cell identifiers and cell type labels, use the <code>cell_id_col</code> and <code>cell_type_col</code> parameters, respectively.<br></p>
<p>As this function uses the <code>SimBu</code> python environment to read the h5ad files and extract the data, it may take some more time to initalize the conda environment at the <em>first</em> usage only.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># example h5ad file, where cell type info is stored in `obs` layer</span>
<span class="va">h5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'anndata.h5ad'</span>, package<span class="op">=</span><span class="st">'SimBu'</span><span class="op">)</span>
<span class="va">ds_h5ad</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/dataset_h5ad.html">dataset_h5ad</a></span><span class="op">(</span>h5ad_file_counts <span class="op">=</span> <span class="va">h5</span>,
                               name <span class="op">=</span> <span class="st">"h5ad_dataset"</span>,
                               cell_id_col <span class="op">=</span> <span class="st">'ID'</span>,                  <span class="co"># this will use the 'ID' column of the metadata as cell identifiers</span>
                               cell_type_col <span class="op">=</span> <span class="st">'cell_type'</span>,         <span class="co"># this will use the 'cell_type' column of the metadata as cell type info</span>
                               cells_in_obs <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>                 <span class="co"># in case your cell information is stored in the var layer, switch to FALSE</span>
<span class="co">#&gt; Filtering genes...</span>
<span class="co">#&gt; Created dataset.</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="merging-datasets">Merging datasets<a class="anchor" aria-label="anchor" href="#merging-datasets"></a>
</h3>
<p>You are able to merge multiple datasets by using the <code>dataset_merge</code> function:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/dataset.html">dataset</a></span><span class="op">(</span>annotation <span class="op">=</span> <span class="va">annotation</span>,
                     count_matrix <span class="op">=</span> <span class="va">counts</span>,
                     tpm_matrix <span class="op">=</span> <span class="va">tpm</span>,
                     name <span class="op">=</span> <span class="st">"test_dataset"</span><span class="op">)</span>
<span class="co">#&gt; Filtering genes...</span>
<span class="co">#&gt; Created dataset.</span>

<span class="va">ds_multiple</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/dataset_merge.html">dataset_merge</a></span><span class="op">(</span>dataset_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ds_seurat</span>, <span class="va">ds</span><span class="op">)</span>,
                                    name <span class="op">=</span> <span class="st">"ds_multiple"</span><span class="op">)</span>
<span class="co">#&gt; Filtering genes...</span>
<span class="co">#&gt; Created dataset.</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h2>
<p>The <code>simulation</code> object contains three named entries:<br></p>
<ul>
<li>
<code>bulk</code>: a SummarizedExperiment object with the pseudo-bulk dataset(s) stored in the <code>assays</code>. They can be accessed like this:</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">simulation</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_bulk.html">simulate_bulk</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ds_multiple</span>,
                                   scenario <span class="op">=</span> <span class="st">"random"</span>, 
                                   scaling_factor <span class="op">=</span> <span class="st">"NONE"</span>,
                                   nsamples <span class="op">=</span> <span class="fl">10</span>, ncells <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="co">#&gt; Finished simulation.</span>

<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">simulation</span><span class="op">$</span><span class="va">bulk</span><span class="op">)</span><span class="op">[[</span><span class="st">"bulk_counts"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; [1] 1000   10</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">simulation</span><span class="op">$</span><span class="va">bulk</span><span class="op">)</span><span class="op">[[</span><span class="st">"bulk_tpm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; [1] 1000   10</span></code></pre></div>
<p>If only the count matrix was given to the dataset initially, only the <code>bulk_counts</code> assay is filled.</p>
<ul>
<li><p><code>cell_fractions</code>: a table where rows represent the simulated samples and columns represent the different simulated cell-types. The entries in the table store the specific cell-type fraction per sample.<br></p></li>
<li><p><code>scaling_vector</code>: a named list, with the used scaling value for each cell from the single cell dataset.<br></p></li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.0 (2021-05-18)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.6 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] curl_4.3.2                  SeuratObject_4.0.4          Seurat_4.1.0                SimBu_0.99.0               </span>
<span class="co">#&gt;  [5] SummarizedExperiment_1.22.0 Biobase_2.52.0              GenomicRanges_1.44.0        GenomeInfoDb_1.28.4        </span>
<span class="co">#&gt;  [9] IRanges_2.26.0              S4Vectors_0.30.2            BiocGenerics_0.38.0         MatrixGenerics_1.4.3       </span>
<span class="co">#&gt; [13] matrixStats_0.61.0          Matrix_1.3-4               </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] utf8_1.2.2              reticulate_1.24         RUnit_0.4.32            tidyselect_1.1.2        htmlwidgets_1.5.4      </span>
<span class="co">#&gt;   [6] grid_4.1.0              BiocParallel_1.26.2     Rtsne_0.15              munsell_0.5.0           codetools_0.2-18       </span>
<span class="co">#&gt;  [11] ica_1.0-2               future_1.24.0           miniUI_0.1.1.1          withr_2.5.0             spatstat.random_2.1-0  </span>
<span class="co">#&gt;  [16] colorspace_2.0-3        filelock_1.0.2          highr_0.9               knitr_1.33              rstudioapi_0.13        </span>
<span class="co">#&gt;  [21] ROCR_1.0-11             tensor_1.5              listenv_0.8.0           labeling_0.4.2          optparse_1.7.1         </span>
<span class="co">#&gt;  [26] GenomeInfoDbData_1.2.6  polyclip_1.10-0         farver_2.1.0            rprojroot_2.0.2         basilisk_1.4.0         </span>
<span class="co">#&gt;  [31] parallelly_1.30.0       vctrs_0.3.8             generics_0.1.2          xfun_0.30               R6_2.5.1               </span>
<span class="co">#&gt;  [36] bitops_1.0-7            spatstat.utils_2.3-0    cachem_1.0.6            DelayedArray_0.18.0     assertthat_0.2.1       </span>
<span class="co">#&gt;  [41] promises_1.2.0.1        scales_1.1.1            gtable_0.3.0            downlit_0.4.0           biocViews_1.60.0       </span>
<span class="co">#&gt;  [46] globals_0.14.0          processx_3.5.3          goftest_1.2-3           rlang_1.0.2             splines_4.1.0          </span>
<span class="co">#&gt;  [51] lazyeval_0.2.2          spatstat.geom_2.3-2     BiocManager_1.30.16     yaml_2.3.5              reshape2_1.4.4         </span>
<span class="co">#&gt;  [56] abind_1.4-5             httpuv_1.6.5            RBGL_1.68.0             tools_4.1.0             ggplot2_3.3.5          </span>
<span class="co">#&gt;  [61] ellipsis_0.3.2          spatstat.core_2.4-0     RColorBrewer_1.1-2      ggridges_0.5.3          Rcpp_1.0.8.3           </span>
<span class="co">#&gt;  [66] plyr_1.8.7              sparseMatrixStats_1.4.2 zlibbioc_1.38.0         purrr_0.3.4             RCurl_1.98-1.6         </span>
<span class="co">#&gt;  [71] basilisk.utils_1.4.0    ps_1.6.0                rpart_4.1-15            deldir_1.0-6            pbapply_1.5-0          </span>
<span class="co">#&gt;  [76] cowplot_1.1.1           zoo_1.8-9               ggrepel_0.9.1           cluster_2.1.2           fs_1.5.2               </span>
<span class="co">#&gt;  [81] here_1.0.1              magrittr_2.0.1          data.table_1.14.2       scattermore_0.8         lmtest_0.9-40          </span>
<span class="co">#&gt;  [86] RANN_2.6.1              fitdistrplus_1.1-8      patchwork_1.1.1         mime_0.10               evaluate_0.14          </span>
<span class="co">#&gt;  [91] xtable_1.8-4            XML_4.0-0               gridExtra_2.3           testthat_3.1.2          compiler_4.1.0         </span>
<span class="co">#&gt;  [96] tibble_3.1.6            KernSmooth_2.23-20      crayon_1.5.1            htmltools_0.5.2         proxyC_0.2.4           </span>
<span class="co">#&gt; [101] mgcv_1.8-36             later_1.3.0             tidyr_1.2.0             RcppParallel_5.1.5      DBI_1.1.2              </span>
<span class="co">#&gt; [106] MASS_7.3-54             getopt_1.20.3           brio_1.1.3              cli_3.2.0               igraph_1.2.11          </span>
<span class="co">#&gt; [111] pkgconfig_2.0.3         pkgdown_2.0.2           dir.expiry_1.0.0        plotly_4.10.0           spatstat.sparse_2.1-0  </span>
<span class="co">#&gt; [116] xml2_1.3.3              stringdist_0.9.8        XVector_0.32.0          BiocCheck_1.28.0        stringr_1.4.0          </span>
<span class="co">#&gt; [121] callr_3.7.0             digest_0.6.27           sctransform_0.3.3       RcppAnnoy_0.0.19        graph_1.70.0           </span>
<span class="co">#&gt; [126] spatstat.data_2.1-2     rmarkdown_2.13          leiden_0.3.9            uwot_0.1.11             shiny_1.7.1            </span>
<span class="co">#&gt; [131] commonmark_1.8.0        lifecycle_1.0.1         nlme_3.1-152            jsonlite_1.7.2          desc_1.4.1             </span>
<span class="co">#&gt; [136] viridisLite_0.4.0       fansi_1.0.3             pillar_1.7.0            lattice_0.20-44         fastmap_1.1.0          </span>
<span class="co">#&gt; [141] httr_1.4.2              survival_3.2-11         glue_1.6.2              png_0.1-7               stringi_1.6.1          </span>
<span class="co">#&gt; [146] memoise_2.0.1           dplyr_1.0.8             irlba_2.3.5             future.apply_1.8.1</span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Alexander Dietrich.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
