<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SimBu">
<title>Getting started with SimBu • SimBu</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with SimBu">
<meta property="og:description" content="SimBu">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SimBu</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.15</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/SimBu.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/sfaira_vignette.html">Public Data Integration using Sfaira</a>
    <a class="dropdown-item" href="../articles/simulator_input_output.html">Inputs and Outputs</a>
    <a class="dropdown-item" href="../articles/simulator_scaling_factors.html">Introducing mRNA bias into simulations with scaling factors</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/omnideconv/SimBu/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Getting started with SimBu</h1>
                        <h4 data-toc-skip class="author">Alexander Dietrich</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/omnideconv/SimBu/blob/HEAD/vignettes/SimBu.Rmd" class="external-link"><code>vignettes/SimBu.Rmd</code></a></small>
      <div class="d-none name"><code>SimBu.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the developmental version of the package, run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"omnideconv/SimBu"</span><span class="op">)</span></span></code></pre></div>
<p>To install from Bioconductor:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"SimBu"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/omnideconv/SimBu" class="external-link">SimBu</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>As complex tissues are typically composed of various cell types, deconvolution tools have been developed to computationally infer their cellular composition from bulk RNA sequencing (RNA-seq) data. To comprehensively assess deconvolution performance, gold-standard datasets are indispensable. Gold-standard, experimental techniques like flow cytometry or immunohistochemistry are resource-intensive and cannot be systematically applied to the numerous cell types and tissues profiled with high-throughput transcriptomics. The simulation of ‘pseudo-bulk’ data, generated by aggregating single-cell RNA-seq (scRNA-seq) expression profiles in pre-defined proportions, offers a scalable and cost-effective alternative. This makes it feasible to create in silico gold standards that allow fine-grained control of cell-type fractions not conceivable in an experimental setup. However, at present, no simulation software for generating pseudo-bulk RNA-seq data exists.<br>
SimBu was developed to simulate pseudo-bulk samples based on various simulation scenarios, designed to test specific features of deconvolution methods. A unique feature of SimBu is the modelling of cell-type-specific mRNA bias using experimentally-derived or data-driven scaling factors. Here, we show that SimBu can generate realistic pseudo-bulk data, recapitulating the biological and statistical features of real RNA-seq data. Finally, we illustrate the impact of mRNA bias on the evaluation of deconvolution tools and provide recommendations for the selection of suitable methods for estimating mRNA content.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>This chapter covers all you need to know to quickly simulate some pseudo-bulk samples!<br></p>
<p>This package can simulate samples from local or public data. This vignette will work with artificially generated data as it serves as an overview for the features implemented in SimBu. For the public data integration using sfaira <span class="citation">(<a href="#ref-Fischer2020" role="doc-biblioref">Fischer et al. 2020</a>)</span>, please refer to the <a href="sfaira_vignette.html">“Public Data Integration”</a> vignette.<br></p>
<p>We will create some toy data to use for our simulations; two matrices with 300 cells each and 1000 genes/features. One represents raw count data, while the other matrix represents scaled TPM-like data. We will assign these cells to some immune cell types.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html" class="external-link">Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">3e5</span>, <span class="fl">5</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">300</span><span class="op">)</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">tpm</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html" class="external-link">Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">3e5</span>, <span class="fl">5</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">300</span><span class="op">)</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">tpm</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="fl">1e6</span> <span class="op">*</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span> <span class="op">/</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">300</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">300</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tpm</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">"ID"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">300</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="st">"cell_type"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"T cells CD4"</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"T cells CD8"</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Macrophages"</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"NK cells"</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"B cells"</span>, <span class="fl">70</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Monocytes"</span>, <span class="fl">20</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="creating-a-dataset">Creating a dataset<a class="anchor" aria-label="anchor" href="#creating-a-dataset"></a>
</h3>
<p>SimBu uses the <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html" class="external-link">SummarizedExperiment</a> class as storage for count data as well as annotation data. Currently it is possible to store two matrices at the same time: raw counts and TPM-like data (this can also be some other scaled count matrix, such as RPKM, but we recommend to use TPMs). These two matrices have to have the same dimensions and have to contain the same genes and cells. Providing the raw count data is <em>mandatory</em>!<br>
SimBu scales the matrix that is added via the <code>tpm_matrix</code> slot by default to 1e6 per cell, if you do not want this, you can switch it off by setting the <code>scale_tpm</code> parameter to <code>FALSE</code>. Additionally, the cell type annotation of the cells has to be given in a dataframe, which has to include the two columns <code>ID</code> and <code>cell_type</code>. If additional columns from this annotation should be transferred to the dataset, simply give the names of them in the <code>additional_cols</code> parameter.<br></p>
<p>To generate a dataset that can be used in SimBu, you can use the <code><a href="../reference/dataset.html">dataset()</a></code> method; other methods exist as well, which are covered in the <a href="simulator_input_output.html">“Inputs &amp; Outputs”</a> vignette.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/dataset.html">dataset</a></span><span class="op">(</span></span>
<span>  annotation <span class="op">=</span> <span class="va">annotation</span>,</span>
<span>  count_matrix <span class="op">=</span> <span class="va">counts</span>,</span>
<span>  tpm_matrix <span class="op">=</span> <span class="va">tpm</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"test_dataset"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Filtering genes...</span></span>
<span><span class="co">#&gt; Created dataset.</span></span></code></pre></div>
<p>SimBu offers basic filtering options for your dataset, which you can apply during dataset generation:<br></p>
<ul>
<li><p><em>filter_genes</em>: if TRUE, genes which have expression values of 0 in all cells will be removed.<br></p></li>
<li><p><em>variance_cutoff</em>: remove all genes with a expression variance below the chosen cutoff.<br></p></li>
<li><p><em>type_abundance_cutoff</em>: remove all cells, which belong to a cell type that appears less the the given amount.<br></p></li>
</ul>
</div>
<div class="section level3">
<h3 id="simulate-pseudo-bulk-datasets">Simulate pseudo bulk datasets<a class="anchor" aria-label="anchor" href="#simulate-pseudo-bulk-datasets"></a>
</h3>
<p>We are now ready to simulate the first pseudo bulk samples with the created dataset:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simulation</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_bulk.html">simulate_bulk</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">ds</span>,</span>
<span>  scenario <span class="op">=</span> <span class="st">"random"</span>,</span>
<span>  scaling_factor <span class="op">=</span> <span class="st">"NONE"</span>,</span>
<span>  ncells <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  nsamples <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>, <span class="co"># this will use 4 threads to run the simulation</span></span>
<span>  run_parallel <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="co"># multi-threading to TRUE</span></span>
<span><span class="co">#&gt; Using parallel generation of simulations.</span></span>
<span><span class="co">#&gt; Finished simulation.</span></span></code></pre></div>
<p><code>ncells</code> sets the number of cells in each sample, while <code>nsamples</code> sets the total amount of simulated samples.<br>
If you want to simulate a specific sequencing depth in your simulations, you can use the <code>total_read_counts</code> parameter to do so. Note that this parameter is only applied on the counts matrix (if supplied), as TPMs will be scaled to 1e6 by default.<br></p>
<p>SimBu can add mRNA bias by using different scaling factors to the simulations using the <code>scaling_factor</code> parameter. A detailed explanation can be found in the <a href="simulator_scaling_factors.html">“Scaling factor”</a> vignette.<br></p>
<p>Currently there are 6 <code>scenarios</code> implemented in the package:<br></p>
<ul>
<li><p><em>even</em>: this creates samples, where all existing cell-types in the dataset appear in the same proportions. So using a dataset with 3 cell-types, this will simulate samples, where all cell-type fractions are 1/3. In order to still have a slight variation between cell type fractions, you can increase the <code>balance_uniform_mirror_scenario</code> parameter (default to 0.01). Setting it to 0 will generate simulations with exactly the same cell type fractions.<br></p></li>
<li><p><em>random</em>: this scenario will create random cell type fractions using all present types for each sample. The random sampling is based on the uniform distribution.<br></p></li>
<li><p><em>mirror_db</em>: this scenario will mirror the exact fractions of cell types which are present in the provided dataset. If it consists of 20% T cells, 30% B cells and 50% NK cells, all simulated samples will mirror these fractions. Similar to the uniform scenario, you can add a small variation to these fractions with the <code>balance_uniform_mirror_scenario</code> parameter.</p></li>
<li><p><em>weighted</em>: here you need to set two additional parameters for the <code><a href="../reference/simulate_bulk.html">simulate_bulk()</a></code> function: <code>weighted_cell_type</code> sets the cell-type you want to be over-representing and <code>weighted_amount</code> sets the fraction of this cell-type. You could for example use <code>B-cell</code> and <code>0.5</code> to create samples, where 50% are B-cells and the rest is filled randomly with other cell-types.<br></p></li>
<li><p><em>pure</em>: this creates simulations of only one single cell-type. You have to provide the name of this cell-type with the <code>pure_cell_type</code> parameter.</p></li>
<li><p><em>custom</em>: here you are able to create your own set of cell-type fractions. When using this scenario, you additionally need to provide a dataframe in the <code>custom_scenario_data</code> parameter, where each row represents one sample (therefore the number of rows need to match the <code>nsamples</code> parameter). Each column has to represent one cell-type, which also occurs in the dataset and describes the fraction of this cell-type in a sample. The fractions per sample need to sum up to 1. An example can be seen here:</p></li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pure_scenario_dataframe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">"B cells"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.1</span>, <span class="fl">0.5</span>, <span class="fl">0.3</span><span class="op">)</span>,</span>
<span>  <span class="st">"T cells"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.8</span>, <span class="fl">0.2</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>  <span class="st">"NK cells"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.1</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span><span class="op">)</span>,</span>
<span>  row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span>, <span class="st">"sample3"</span>, <span class="st">"sample4"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pure_scenario_dataframe</span></span>
<span><span class="co">#&gt;         B.cells T.cells NK.cells</span></span>
<span><span class="co">#&gt; sample1     0.2     0.3      0.5</span></span>
<span><span class="co">#&gt; sample2     0.1     0.8      0.1</span></span>
<span><span class="co">#&gt; sample3     0.5     0.2      0.3</span></span>
<span><span class="co">#&gt; sample4     0.3     0.5      0.2</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h3>
<p>The <code>simulation</code> object contains three named entries:<br></p>
<ul>
<li>
<code>bulk</code>: a SummarizedExperiment object with the pseudo-bulk dataset(s) stored in the <code>assays</code>. They can be accessed like this:</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">simulation</span><span class="op">$</span><span class="va">bulk</span><span class="op">)</span><span class="op">[[</span><span class="st">"bulk_counts"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1000   10</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">simulation</span><span class="op">$</span><span class="va">bulk</span><span class="op">)</span><span class="op">[[</span><span class="st">"bulk_tpm"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1000   10</span></span></code></pre></div>
<p>If only a single matrix was given to the dataset initially, only one assay is filled.</p>
<ul>
<li><p><code>cell_fractions</code>: a table where rows represent the simulated samples and columns represent the different simulated cell-types. The entries in the table store the specific cell-type fraction per sample.<br></p></li>
<li><p><code>scaling_vector</code>: a named list, with the used scaling value for each cell from the single cell dataset.<br></p></li>
</ul>
<p>It is also possible to merge simulations:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simulation2</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_bulk.html">simulate_bulk</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">ds</span>,</span>
<span>  scenario <span class="op">=</span> <span class="st">"even"</span>,</span>
<span>  scaling_factor <span class="op">=</span> <span class="st">"NONE"</span>,</span>
<span>  ncells <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  nsamples <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,</span>
<span>  run_parallel <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Using parallel generation of simulations.</span></span>
<span><span class="co">#&gt; Finished simulation.</span></span>
<span><span class="va">merged_simulations</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/merge_simulations.html">merge_simulations</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">simulation</span>, <span class="va">simulation2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Finally here is a barplot of the resulting simulation:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_simulation.html">plot_simulation</a></span><span class="op">(</span>simulation <span class="op">=</span> <span class="va">merged_simulations</span><span class="op">)</span></span></code></pre></div>
<p><img src="SimBu_files/figure-html/unnamed-chunk-10-1.png" width="768"></p>
</div>
</div>
<div class="section level2">
<h2 id="more-features">More features<a class="anchor" aria-label="anchor" href="#more-features"></a>
</h2>
<div class="section level3">
<h3 id="simulate-using-a-whitelist-and-blacklist-of-cell-types">Simulate using a whitelist (and blacklist) of cell-types<a class="anchor" aria-label="anchor" href="#simulate-using-a-whitelist-and-blacklist-of-cell-types"></a>
</h3>
<p>Sometimes, you are only interested in specific cell-types (for example T cells), but the dataset you are using has too many other cell-types; you can handle this issue during simulation using the <code>whitelist</code> parameter:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simulation</span> <span class="op">&lt;-</span> <span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_bulk.html">simulate_bulk</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">ds</span>,</span>
<span>  scenario <span class="op">=</span> <span class="st">"random"</span>,</span>
<span>  scaling_factor <span class="op">=</span> <span class="st">"NONE"</span>,</span>
<span>  ncells <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  nsamples <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,</span>
<span>  run_parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  whitelist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T cells CD4"</span>, <span class="st">"T cells CD8"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Using parallel generation of simulations.</span></span>
<span><span class="co">#&gt; Finished simulation.</span></span>
<span><span class="fu">SimBu</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_simulation.html">plot_simulation</a></span><span class="op">(</span>simulation <span class="op">=</span> <span class="va">simulation</span><span class="op">)</span></span></code></pre></div>
<p><img src="SimBu_files/figure-html/unnamed-chunk-11-1.png" width="768"></p>
<p>In the same way, you can also provide a <code>blacklist</code> parameter, where you name the cell-types you <strong>don’t</strong> want to be included in your simulation.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 20.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] SimBu_0.99.15</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] MatrixGenerics_1.8.1        Biobase_2.56.0             </span></span>
<span><span class="co">#&gt;  [3] sass_0.4.2                  tidyr_1.2.1                </span></span>
<span><span class="co">#&gt;  [5] jsonlite_1.8.2              bslib_0.4.0                </span></span>
<span><span class="co">#&gt;  [7] RcppParallel_5.1.5          highr_0.9                  </span></span>
<span><span class="co">#&gt;  [9] stats4_4.2.1                GenomeInfoDbData_1.2.8     </span></span>
<span><span class="co">#&gt; [11] yaml_2.3.6                  pillar_1.8.1               </span></span>
<span><span class="co">#&gt; [13] lattice_0.20-45             glue_1.6.2                 </span></span>
<span><span class="co">#&gt; [15] digest_0.6.30               GenomicRanges_1.48.0       </span></span>
<span><span class="co">#&gt; [17] RColorBrewer_1.1-3          XVector_0.36.0             </span></span>
<span><span class="co">#&gt; [19] colorspace_2.0-3            htmltools_0.5.3            </span></span>
<span><span class="co">#&gt; [21] Matrix_1.5-1                pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt; [23] zlibbioc_1.42.0             purrr_0.3.5                </span></span>
<span><span class="co">#&gt; [25] scales_1.2.1                BiocParallel_1.30.4        </span></span>
<span><span class="co">#&gt; [27] tibble_3.1.8                generics_0.1.3             </span></span>
<span><span class="co">#&gt; [29] farver_2.1.1                IRanges_2.30.1             </span></span>
<span><span class="co">#&gt; [31] ggplot2_3.3.6               ellipsis_0.3.2             </span></span>
<span><span class="co">#&gt; [33] cachem_1.0.6                withr_2.5.0                </span></span>
<span><span class="co">#&gt; [35] SummarizedExperiment_1.26.1 BiocGenerics_0.42.0        </span></span>
<span><span class="co">#&gt; [37] cli_3.4.1                   magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [39] memoise_2.0.1               evaluate_0.17              </span></span>
<span><span class="co">#&gt; [41] fs_1.5.2                    fansi_1.0.3                </span></span>
<span><span class="co">#&gt; [43] textshaping_0.3.6           tools_4.2.1                </span></span>
<span><span class="co">#&gt; [45] data.table_1.14.4           lifecycle_1.0.3            </span></span>
<span><span class="co">#&gt; [47] matrixStats_0.62.0          stringr_1.4.1              </span></span>
<span><span class="co">#&gt; [49] S4Vectors_0.34.0            munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [51] DelayedArray_0.22.0         compiler_4.2.1             </span></span>
<span><span class="co">#&gt; [53] pkgdown_2.0.6               jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [55] GenomeInfoDb_1.32.4         proxyC_0.3.3               </span></span>
<span><span class="co">#&gt; [57] systemfonts_1.0.4           rlang_1.0.6                </span></span>
<span><span class="co">#&gt; [59] grid_4.2.1                  RCurl_1.98-1.9             </span></span>
<span><span class="co">#&gt; [61] bitops_1.0-7                labeling_0.4.2             </span></span>
<span><span class="co">#&gt; [63] rmarkdown_2.17              gtable_0.3.1               </span></span>
<span><span class="co">#&gt; [65] codetools_0.2-18            R6_2.5.1                   </span></span>
<span><span class="co">#&gt; [67] knitr_1.40                  dplyr_1.0.10               </span></span>
<span><span class="co">#&gt; [69] fastmap_1.1.0               utf8_1.2.2                 </span></span>
<span><span class="co">#&gt; [71] rprojroot_2.0.3             ragg_1.2.3                 </span></span>
<span><span class="co">#&gt; [73] desc_1.4.2                  stringi_1.7.8              </span></span>
<span><span class="co">#&gt; [75] parallel_4.2.1              Rcpp_1.0.9                 </span></span>
<span><span class="co">#&gt; [77] vctrs_0.4.2                 tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt; [79] xfun_0.34                   sparseMatrixStats_1.8.0</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Fischer2020" class="csl-entry">
Fischer, David S., Leander Dony, Martin König, Abdul Moeed, Luke Zappia, Sophie Tritschler, Olle Holmberg, Hananeh Aliee, and Fabian J. Theis. 2020. <span>“Sfaira Accelerates Data and Model Reuse in Single Cell Genomics.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2020.12.16.419036" class="external-link">https://doi.org/10.1101/2020.12.16.419036</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Alexander Dietrich.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
